doc/005-data-sources.md

```diff
@@ -64,34 +64,69 @@ For more information on configuring the data sets in your data source using `dat
 
 Many REST APIs use a paging mechanism to allow data to be retrieved in chunks. For example, some APIs take a `pageSize` and `pageNumber` argument which specify how many items to return, and the page number to start reading from, respectively.
 
-While it is possible to configure many aspects of your data sets using the `dataSource.json` file, there is currently no mechanism to handle paging. To handle paging properly, you will have to write a bit of code.
-
-To implement paging in a data source that derives from RestDataSourceBase, override the `getDataFromUrl` method:
-
-```
-  protected async getDataFromUrl(
-    dataSetName: string,
-    dataSet: IDataSet,
-    relativeUrl: string,
-    params: DataValueMap
-  ) {
-    // No need for paging if we are just getting one object.
-    if (dataSet.result === 'object' && !dataSet.filter) {
-      return super.getDataFromUrl(dataSetName, dataSet, relativeUrl, params);
-    }
-
-    // TODO: Handle paging for My Service here.
-    const PAGE_SIZE = 50;
-    const results: DataObject[] = [];
-    const pageNumber = 0;
-    let pageResults;
-    do {
-        pageResults = await fetchPage(PAGE_SIZE, pageNumber++)
-        Array.prototype.push.apply(results, pageResults);
-    } while (pageResults.length < PAGE_SIZE)
-  }
+Three paging styles are supported.
+
+1.  __Page Based.__  Begin paging at a starting value and increment the page value by 1 after each iteration.  Return at most `limitValue` items.
+
+```
+"pagingScheme": {
+  "type": "pageBased",
+  "request": {
+    "pageParameter": "page",
+    "pageStartingValue": 1,
+    "limitParameter": "pageSize",
+    "limitValue": 100
+  },
+  "pageUntil": "noDataLeft"
+}
+```
+
+2.  __Offset And Limit.__  Beging paging at a starting value and increment the offset by the number of elements in a full page.  Return at most `limitValue` items.
+
+```
+"pagingScheme": {
+  "type": "offsetAndLimit",
+  "request": {
+    "offsetParameter": "offset",
+    "offsetStartingValue": 0,
+    "limitParameter": "limit",
+    "limitValue": 100
+  },
+  "response": {
+    "totalCount": "pagination.total"
+  },
+  "pageUntil": "reachTotalCount"
+}
 ```
 
+3.  __Next Token.__  Begin paging and continue until `nextToken` is no longer provided.  Return at most `limitValue` items.  Tokens may be a unique string returned from the external service or a url.
```

Comment:
I think you should specify what means what in the example below. Like, what is the meaning of "bookmark" and "next.bookmark"?